### PR TITLE
Implemented Transaction - related to common transaction

### PR DESCRIPTION
**Pull request by sirano11**
## Proposed changes

This PR introduces Common transaction class.
  - AbstractTransaction
  - LegacyTransaction
  - ValueTransfer
  - ValueTransferMemo
  - SmartContractDeploy
  - SmartContractExecution
  - Cancel
  - ChainDataAnchoring

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-java/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-java)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

related #97 

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java [hunk 1]
```diff
@@ -657,4 +652,61 @@ public String getChainId() {
...
+        //"From" field in LegacyTransaction allows null
+        if(this instanceof LegacyTransaction) {
+            if(from == null || from.isEmpty() || from.equals("0x")) from = "0x";
+        } else {
+            if(from == null) {
+                throw new IllegalArgumentException("from is missing.");
+            }
+
+            if(!Utils.isAddress(from)) {
+                throw new IllegalArgumentException("Invalid address.");
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java:670]: Can we add invalid from string in error message?

File: core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java [hunk 2]
```diff
@@ -657,4 +652,61 @@ public String getChainId() {
...
     }
+
+    private void setType(String type) {
+        this.type = type;
+    }
+
+    private void setFrom(String from) {
+        //"From" field in LegacyTransaction allows null
+        if(this instanceof LegacyTransaction) {
+            if(from == null || from.isEmpty() || from.equals("0x")) from = "0x";
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java:663]: isEmpty will return 'true' when String is ""?

File: core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java [hunk 3]
```diff
@@ -657,4 +652,61 @@ public String getChainId() {
...
+
+            if(!Utils.isAddress(from)) {
+                throw new IllegalArgumentException("Invalid address.");
+            }
+        }
+
+        this.from = from;
+    }
+
+    private void setGas(String gas) {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java:677]: Will setGas accept only String type of gas?

File: core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java (hunk 2 above)
Comment (sirano11) [reply core/src/main/java/com/klaytn/caver/transaction/AbstractTransaction.java:663]: Yes. It check if String length has 0

File: core/src/main/java/com/klaytn/caver/transaction/type/ChainDataAnchoring.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    /**
+     * Setter function for input
+     * @param input Data of the service chain.
+     */
+    public void setInput(String input) {
+        if(input == null) {
+            throw new IllegalArgumentException("input is missing.");
+        }
+
+        if(!input.equals("0x") && !Utils.isHex(input)) {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/ChainDataAnchoring.java:208]: Will Utils.isHex function return false with "0x"?

File: core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java [hunk 5]
```diff
@@ -160,7 +161,7 @@ public static LegacyTransaction decode(byte[] rlpEncoded) {
      */
     @Override
     public void appendSignatures(KlaySignatureData signatureData) {
-        if(this.getSignatures().size() != 0) {
+        if(this.getSignatures().size() != 0 && !Utils.isEmptySig(this.getSignatures().get(0))) {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java:164]: Please implement isEmptySig for List<KlaySignatureData> :)

File: core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java [hunk 6]
```diff
@@ -270,4 +271,38 @@ public String getInput() {
...
+        this.input = input;
+    }
+
+    public void setValue(String value) {
+        if(value == null) {
+            throw new IllegalArgumentException("value is missing");
+        }
+
+        if(!Utils.isHex(value)) {
+            throw new IllegalArgumentException("Invalid value.");
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java:304]:
```suggestion
            throw new IllegalArgumentException("Invalid value: " + value);
```

File: core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java [hunk 7]
```diff
@@ -270,4 +271,38 @@ public String getInput() {
...
+        }
+        this.input = input;
+    }
+
+    public void setValue(String value) {
+        if(value == null) {
+            throw new IllegalArgumentException("value is missing");
+        }
+
+        if(!Utils.isHex(value)) {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/LegacyTransaction.java:303]: How about adding this validation or other(isNumber??) to other setter? for example setGasPrice or setNonce.

File: core/src/main/java/com/klaytn/caver/transaction/type/SmartContractDeploy.java [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+     */
+    public String getInput() {
+        return input;
+    }
+
+    /**
+     * Getter function for humanReadable
+     * @return boolean
+     */
+    public boolean isHumanReadable() {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/SmartContractDeploy.java:318]: How about change this getter name to getHumanReadable?

File: core/src/main/java/com/klaytn/caver/transaction/type/SmartContractExecution.java [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+     */
+    private SmartContractExecution(SmartContractExecution.Builder builder) {
+        super(builder);
+        setTo(builder.to);
+        setValue(builder.value);
+        setInput(builder.input);
+    }
+
+    /**
+     * Creates an ValueTransferMemo instance
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/SmartContractExecution.java:86]:
```suggestion
     * Creates a SmartContractExecution instance
```

File: core/src/main/java/com/klaytn/caver/transaction/type/ValueTransfer.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+public class ValueTransfer extends AbstractTransaction {
+    /**
+     * The account address that will receive the transferred value.
+     */
+    String to;
+
+    /**
+     * The amount of KLAY in peb to be transferred.
+     */
+    String value = "0x00";
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/ValueTransfer.java:29]: I think for value should be defined by user with ValueTransfer transaction.
How about define value without default value?

File: core/src/main/java/com/klaytn/caver/transaction/type/ValueTransferMemo.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+    /**
+     * The account address that will receive the transferred value.
+     */
+    String to;
+
+    /**
+     * The amount of KLAY in peb to be transferred.
+     */
+    String value = "0x00";
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/ValueTransferMemo.java:30]: Here value and input also i think should be defined by user.

File: core/src/main/java/com/klaytn/caver/transaction/type/SmartContractExecution.java [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * The amount of KLAY in peb to be transferred.
+     */
+    String value = "0x00";
+
+    /**
+     * Data attached to the transaction, used for transaction execution.
+     */
+    String input = "0x";
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/SmartContractExecution.java:35]: How about remove default value of input?

File: core/src/main/java/com/klaytn/caver/transaction/type/SmartContractDeploy.java [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+
+    /**
+     * The amount of KLAY in peb to be transferred.
+     */
+    String value = "0x00";
+
+    /**
+     * Data attached to the transaction, used for transaction execution.
+     */
+    String input = " 0x";
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/SmartContractDeploy.java:36]: This also i think input should be defined.

File: core/src/main/java/com/klaytn/caver/transaction/type/ValueTransferMemo.java [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+     */
+    public ValueTransferMemo(Builder builder) {
+        super(builder);
+        setTo(builder.to);
+        setValue(builder.value);
+        setInput(builder.input);
+    }
+
+    /**
+     * Creates an ValueTransferMemo instance
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/transaction/type/ValueTransferMemo.java:86]:
```suggestion
     * Creates a ValueTransferMemo instance
```

File: core/src/main/java/com/klaytn/caver/transaction/type/ChainDataAnchoring.java (hunk 4 above)
Comment (sirano11) [reply core/src/main/java/com/klaytn/caver/transaction/type/ChainDataAnchoring.java:208]: Yes, it returns false.

File: core/src/main/java/com/klaytn/caver/utils/Utils.java [hunk 15]
```diff
@@ -126,7 +126,7 @@ public static boolean isEmptySig(KlaySignatureData signatureData) {
     public static boolean isEmptySig(List<KlaySignatureData> signatureDataList) {
         KlaySignatureData emptySig = KlaySignatureData.getEmptySignature();
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/utils/Utils.java:128]: How about adding length check here?

**Pull request merged**